File: tests/ut/cpp/python_input/gtest_input/pre_activate/all_to_all_unify_mindir_test.py
# Copyright 2021 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
import mindspore as ms
from mindspore import context
from mindspore.ops.operations.comm_ops import AlltoAll, NeighborExchange
from mindspore.communication.management import GlobalComm, init

context.set_context(device_target="Ascend")
GlobalComm.CHECK_ENVS = False
init("hccl")
GlobalComm.CHECK_ENVS = True

class FnDict:
    def __init__(self):
        self.fnDict = {}

    def __call__(self, fn):
        self.fnDict[fn.__name__] = fn

    def __getitem__(self, name):
        return self.fnDict[name]

def test_neighbor_exchange(tag):
    fns = FnDict()
    neighbor = NeighborExchange(send_rank_ids=[0], recv_rank_ids=[1], recv_shapes=([2, 2],), send_shapes=([2, 2],),
                                recv_type=ms.float32)
    @fns
    def before(x):
        return neighbor(x)

    return fns[tag]

def test_all_to_all(tag):
    context.set_auto_parallel_context(device_num=2, global_rank=0)
    fns = FnDict()
    altoall = AlltoAll(split_count=2, split_dim=2, concat_dim=3)
    @fns
    def before(x):
        return altoall(x)

    return fns[tag]

File: tests/ut/cpp/pre_activate/ascend/mindir/all_to_all_unify_mindir_test.cc
/**
 * Copyright 2021 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "common/backend_common_test.h"
#include "frontend/operator/ops.h"
#include "include/common/debug/anf_ir_dump.h"
#include "common/py_func_graph_fetcher.h"
#include "include/backend/anf_runtime_algorithm.h"
#include "backend/common/optimizer/optimizer.h"
#include "backend/common/optimizer/pass_manager.h"
#include "backend/common/pass/convert_const_input_to_attr.h"
#include "include/common/utils/utils.h"
#include "utils/ms_utils.h"

namespace mindspore {
namespace opt {
class TestAllToAllUnifyMindIr : public BackendCommon {
 public:
  TestAllToAllUnifyMindIr() : getPyFun_("gtest_input.pre_activate.all_to_all_unify_mindir_test", true) {}
  ~TestAllToAllUnifyMindIr() override = default;

 public:
  UT::PyFuncGraphFetcher getPyFun_;
};

TEST_F(TestAllToAllUnifyMindIr, test_neighbor_exchange) {
  FuncGraphPtr g = getPyFun_.CallAndParseRet("test_neighbor_exchange", "before");
  ASSERT_TRUE(g != nullptr);
  std::vector<int64_t> shp_x{2, 2};
  auto x_abstract = std::make_shared<abstract::AbstractTuple>(
    AbstractBasePtrList{std::make_shared<abstract::AbstractTensor>(kFloat32, shp_x)});
  AbstractBasePtrList args_spec_list{x_abstract};
  auto func_graph = GetKernelGraph(g, args_spec_list);
  ASSERT_TRUE(func_graph != nullptr);
  bool has_all_to_all_v_node = false;
  for (const auto &n : TopoSort(func_graph->get_return())) {
    ASSERT_FALSE(IsPrimitiveCNode(n, prim::kPrimNeighborExchange));
    if (IsPrimitiveCNode(n, prim::kPrimAllToAllv)) {
      has_all_to_all_v_node = true;
    }
  }
  ASSERT_TRUE(has_all_to_all_v_node);
}

TEST_F(TestAllToAllUnifyMindIr, test_all_to_all) {
  FuncGraphPtr g = getPyFun_.CallAndParseRet("test_all_to_all", "before");
  ASSERT_TRUE(g != nullptr);
  std::vector<int64_t> shp_x{4, 2, 224, 224};
  auto x_abstract = std::make_shared<abstract::AbstractTensor>(kFloat32, shp_x);
  AbstractBasePtrList args_spec_list{x_abstract};
  auto func_graph = GetKernelGraph(g, args_spec_list);
  ASSERT_TRUE(func_graph != nullptr);
  bool has_all_to_all_v_node = false;
  bool has_concat_node = false;
  bool has_split_v_node = false;
  for (const auto &n : TopoSort(func_graph->get_return())) {
    ASSERT_FALSE(IsPrimitiveCNode(n, prim::kPrimAllToAll));
    if (IsPrimitiveCNode(n, prim::kPrimAllToAllv)) {
      has_all_to_all_v_node = true;
    }
    if (IsPrimitiveCNode(n, prim::kPrimConcatD)) {
      has_concat_node = true;
    }
    if (IsPrimitiveCNode(n, prim::kPrimSplitVD)) {
      has_split_v_node = true;
    }
  }
  ASSERT_TRUE(has_all_to_all_v_node);
  ASSERT_TRUE(has_concat_node);
  ASSERT_TRUE(has_split_v_node);
}
}  // namespace opt
}  // namespace mindspore
